src/net/http_network.rs

```diff
@@ -0,0 +1,155 @@
+use crate::{
+    helpers::{
+        fabric::{MessageChunks, Network},
+        Error,
+    },
+    net::{
+        discovery::{peer, PeerDiscovery},
+        MpcHelperClient,
+    },
+    protocol::QueryId,
+};
+use futures::{ready, Sink};
+use pin_project::pin_project;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::ReceiverStream;
+use tokio_util::sync::{PollSendError, PollSender};
+
+/// Wrapper around a [`PollSender`] to modify the error message to match what the [`Network`] trait
+/// requires. The only error that [`PollSender`] will generate is "channel closed", and thus is the
+/// only error message forwarded from this [`Sink`].
+#[pin_project]
+pub struct HttpMessageSink<T> {
+    #[pin]
+    inner: PollSender<T>,
+}
+
+impl<T: Send + 'static> HttpMessageSink<T> {
+    #[must_use]
+    pub fn new(sender: mpsc::Sender<T>) -> Self {
+        Self {
+            inner: PollSender::new(sender),
+        }
+    }
+}
+
+impl<T: Send + 'static> Sink<T> for HttpMessageSink<T>
+where
+    Error: From<PollSendError<T>>,
+{
+    type Error = Error;
+
+    fn poll_ready(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        ready!(self.project().inner.poll_ready(cx)?);
+        Poll::Ready(Ok(()))
+    }
+
+    fn start_send(self: Pin<&mut Self>, item: T) -> Result<(), Self::Error> {
+        self.project().inner.start_send(item)?;
+        Ok(())
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        ready!(self.project().inner.poll_flush(cx)?);
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        ready!(self.project().inner.poll_close(cx))?;
+        Poll::Ready(Ok(()))
+    }
+}
+
+/// Http implementation of a [`Network`]. Uses channels for both [`Sink`] and [`Stream`]
+/// implementations.
+/// # Panics
+/// if `recv_stream` or `recv_messages` called more than once.
+#[allow(dead_code)] // TODO: WIP
+pub struct HttpNetwork<'a> {
```

Comment:
is it going to stay for streaming model or just temporary to get discrete working?

Reply:
even under streaming model we will need an implementation of the `Network` trait, right? or are you talking about something else?

Reply:
we would need `Network` implementation for sure. I wonder if this implementation can be used in streaming model. The reason I am asking is because again this looks very similar to `InMemoryEndpoint`. If we need it in the long term (which does not seem to be the case, but lmk if I am wrong), we could invest in keeping one implementation

Reply:
and also name is a bit suspicious - nothing that is related to HTTP is here

Reply:
if you look at this PR: https://github.com/private-attribution/ipa/pull/221

it starts using the HTTP clients

Reply:
to answer your question about moving to Streaming, the network trait itself will change to accommodate the new api that includes `ChannelId`, so I figure at that time we can modify this `HttpNetwork` to adapt to new API, and to handle streams

Reply:
ack, now it makes sense